yew/src/virtual_dom/vtag.rs

```diff
@@ -169,21 +169,19 @@ impl VTag {
     }
 
     /// Adds attribute to a virtual node. Not every attribute works when
-    /// it set as attribute. We use workarounds for:
+    /// it is set as an attribute. We use workarounds for:
     /// `type/kind`, `value` and `checked`.
     ///
     /// If this virtual node has this attribute present, the value is replaced.
-    pub fn add_attribute<T: ToString>(&mut self, name: &str, value: &T) {
-        self.attributes.insert(name.to_owned(), value.to_string());
+    pub fn add_attribute(&mut self, name: &'static str, value: impl Into<Cow<'static, str>>) {
+        self.attributes.push((name, value.into()));
```

Comment:
>With this approach, there is no guarantee that the same attribute can only be added once. If the same attribute is already present, the new value will still be appended to the attributes list. This could be problematic. Imagine when you diff two nodes, only the earlier attribute has changed. In this case, Yew would update the DOM with an attribute value that should have been ignored.

The `html!` macro protects from that. The big difference is that one is a runtime and the other a compile time check. It would be advantageous to leverage the zero runtime cost checks of `html!` that most use and expect the people that integrate directly with the VTag API to know what they are doing. 

>I personally would prefer if we used an IndexMap. It seems to be the best of both worlds, because it enforces unique attributes, has quick lookup, and has ordered iteration for the most common case. What do you think?

An IndexMap needs to be parsed from a string and requires heap allocations. Those should be avoided, if possible. This simple example demonstrates the overhead https://godbolt.org/z/PefvGY of only heap allocations. Now imagine that overhead multiplied by the amount of attributes in a large tree diff. Hardly negligible. IndexMap does offer some protection against incorrect use but only for people not using the `html!` macro and at a cost for everyone.

Reply:
> It would be advantageous to leverage the zero runtime cost checks of html! that most use and expect the people that integrate directly with the VTag API to know what they are doing.

Yeah, this is a good point. I think that's a good mental model when approaching tradeoffs, the macro can handle quite a bit at compile time. However, I think that the `VTag` should be dev friendly when devs reach for it. An append only API isn't very great. We could directly expose the `attributes` field to get around this.

> An IndexMap needs to be constructed and requires heap allocations

I'm not sure what you're getting at here. `Vec` also requires heap allocations so I think the difference in memory allocation will be negligible.

---

To summarize, I still think `IndexMap` is a great compromise because it will help optimize for the most common case (same attributes in order) without making the `VTag` API more error-prone.

Reply:
>I'm not sure what you're getting at here. Vec also requires heap allocations so I think the difference in memory allocation will be negligible.

Sorry, I wasn't very clear. And `IndexedMap` needs to be constructed from a vector, so that is extra heap allocations. Also, because an `IndexedMap` needs to deduplicate at runtime it creates more data conflicts during construction than a `Vec`, reducing parallelism. Also requires hashing. 

>Yeah, this is a good point. I think that's a good mental model when approaching tradeoffs, the macro can handle quite a bit at compile time. However, I think that the VTag should be dev friendly when devs reach for it. An append only API isn't very great. We could directly expose the attributes field to get around this.

First of all, I suggest using a `HashMap` instead of an `IndexedMap` (as it is in the master branch) because it is slightly faster and deduplicated attribute order is purely cosmetic. 

How about the best of both worlds and store the attributes as 
```rust
enum Attributes {
  Vec(Vec<_, _>),
  Map(HashMap<_, _>),
}
```
on the `VTag`? The field will be public and also more conveniently exposed to developers via a 
```rust 
fn attributes(&mut self) -> &mut HashMap<_, _>
```
method that automatically converts any stored `Vec` into `HashMap`.

That way we can both leverage the compile compile-time checks of the `html` macro and provide a convenient API for developers. What do you think?

Reply:
Sounds good to me! That is really the best of both worlds. Let's use `IndexMap` instead of `HashMap` because of the 50% perf improvement.

Some API suggestions:
- `fn attributes` can be named `fn attributes_mut`
- `fn add_attributes` should be renamed to `fn set_attributes` and override any current attributes. It can take a `Into<Attributes>` argument
- Remove `add_attribute` method

Looking forward to hearing your thoughts

Reply:
@jstarry When modifying tests and examples I found that not having `add_attribute` is inconvenient. I suggest keeping `add_attribute()` as a shorthand for `tag.attributes_mut().insert(...)` but also adding `push_attribute()` for DRYing code in the `html!` macro.

Reply:
Sounds good to me